crates/ruff_linter/src/rules/pycodestyle/rules/logical_lines/blank_lines.rs

```diff
@@ -0,0 +1,620 @@
+/// - [Flake 8 rule](https://www.flake8rules.com/rules/E305.html)
+#[violation]
+pub struct BlankLinesAfterFunctionOrClass(pub u32);
+
+impl AlwaysFixableViolation for BlankLinesAfterFunctionOrClass {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let BlankLinesAfterFunctionOrClass(blank_lines) = self;
+        format!("expected 2 blank lines after class or function definition, found ({blank_lines})")
+    }
+
+    fn fix_title(&self) -> String {
+        "Add missing blank line(s)".to_string()
+    }
+}
+
+/// ## What it does
+/// Checks for for 1 blank line between nested functions/classes definitions.
+///
+/// ## Why is this bad?
+/// PEP 8 recommends the using blank lines as following:
+/// - Two blank lines are expected between functions and classes
+/// - One blank line is expected between methods of a class.
+///
+/// ## Example
+/// ```python
+/// def outer():
+///     def inner():
+///         pass
+///     def inner2():
+///         pass
+/// ```
+///
+/// Use instead:
+/// ```python
+/// def outer():
+///     def inner():
+///         pass
+///
+///     def inner2():
+///         pass
+/// ```
+///
+/// ## References
+/// - [PEP 8](https://peps.python.org/pep-0008/#blank-lines)
+/// - [Flake 8 rule](https://www.flake8rules.com/rules/E306.html)
+#[violation]
+pub struct BlankLinesBeforeNestedDefinition(pub u32);
+
+impl AlwaysFixableViolation for BlankLinesBeforeNestedDefinition {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let BlankLinesBeforeNestedDefinition(blank_lines) = self;
+        format!("Expected 1 blank line before a nested definition, found {blank_lines}")
+    }
+
+    fn fix_title(&self) -> String {
+        "Add missing blank line".to_string()
+    }
+}
+
+/// Returns `true` if line is a docstring only line.
+fn is_docstring(line: Option<&LogicalLine>) -> bool {
+    line.is_some_and(|line| {
+        !line.tokens_trimmed().is_empty()
+            && line
+                .tokens_trimmed()
+                .iter()
+                .all(|token| matches!(token.kind(), TokenKind::String))
+    })
+}
+
+/// Returns `true` if the token is Async, Class or Def
+fn is_top_level_token(token: Option<TokenKind>) -> bool {
+    token.is_some_and(|token| matches!(token, TokenKind::Class | TokenKind::Def | TokenKind::Async))
+}
+
+/// Returns `true` if the token is At, Async, Class or Def
+fn is_top_level_token_or_decorator(token: TokenKind) -> bool {
+    matches!(
+        token,
+        TokenKind::Class | TokenKind::Def | TokenKind::Async | TokenKind::At
+    )
+}
+
+/// E301, E302, E303, E304, E305, E306
+#[allow(clippy::too_many_arguments)]
+pub(crate) fn blank_lines(
+    line: &LogicalLine,
+    tracked_vars: &mut BlankLinesTrackingVars,
+    prev_indent_level: Option<usize>,
+    indent_level: usize,
+    indent_size: usize,
+    locator: &Locator,
+    stylist: &Stylist,
+    context: &mut LogicalLinesContext,
+) {
```

Comment:
Nit: We could implement this as a method on `BlankLinesTrackingVars` instead (and rename the struct to `BlankLinesChecker`

```rust
impl BlankLinesChecker {
	pub(crate) fn check_line(&mut self, line: &LogicalLine, prev_indent_level: Option<usize>, indent_level: usize, ...) {
		...
	}
}
```

and you would use it

```
let mut blank_lines_checker = BlankLinesChecker::default();
...

for line in logical_lines {
	blank_lines_checker.check_line(line, ...);
}
```